module/rdpCursor.c

```diff
@@ -228,54 +228,110 @@ rdpSpriteSetCursorCon(rdpClientCon *clientCon,
                       DeviceIntPtr pDev, ScreenPtr pScr, CursorPtr pCurs,
                       int x, int y)
 {
-    uint8_t cur_data[32 * (32 * 4)];
-    uint8_t cur_mask[32 * (32 / 8)];
+    uint8_t *cur_data;
+    uint8_t *cur_mask;
     uint8_t *mask;
     uint8_t *data;
-    int i;
-    int j;
-    int w;
-    int h;
-    int p;
+    int index;
+    int jndex;
+    int server_width;
+    int server_height;
+    int pixel;
     int xhot;
     int yhot;
     int paddedRowBytes;
     int fgcolor;
     int bgcolor;
-    int bpp;
+    int client_max_width;
+    int client_max_height;
+    int sending_width;
+    int sending_height;
+    int sending_bpp;
+    int can_do_new;
+    int can_do_large;
 
     LLOGLN(10, ("rdpSpriteSetCursorCon:"));
     if (clientCon->suppress_output)
     {
         LLOGLN(10, ("rdpSpriteSetCursorCon: suppress_output set"));
         return;
     }
-    w = pCurs->bits->width;
-    h = pCurs->bits->height;
-    if ((pCurs->bits->argb != 0) &&
-        (clientCon->client_info.pointer_flags & 1))
+    if (pCurs == NULL)
     {
-        bpp = 32;
-        paddedRowBytes = PixmapBytePad(w, 32);
+        return;
+    }
+    if (pCurs->bits == NULL)
+    {
+        return;
+    }
+    if (clientCon->client_info.size == 0)
+    {
+        return;
+    }
+    cur_data = (uint8_t *)calloc(1, 96 * 96 * 4 + 96 * 96 / 8);
+    if (cur_data == NULL)
+    {
+        return;
+    }
+    cur_mask = cur_data + 96 * 96 * 4;
+    client_max_width = 32;
+    client_max_height = 32;
+    sending_bpp = 0;
+    can_do_new = clientCon->client_info.pointer_flags & 1;
+    can_do_large = 0;
+#if defined(CLIENT_INFO_CURRENT_VERSION)
+#if CLIENT_INFO_CURRENT_VERSION >= 20230423
+    can_do_large = clientCon->client_info.large_pointer_support_flags & 1;
```

Comment:
If you tested `clientCon->cursor_shmemptr` here, you could maybe fall back to 32x32 if the `shmat()` had failed for some reason. This might be preferable to sending no cursor at all.

Reply:
Good point, I'll add that check.  Does it looks like I used CLIENT_INFO_CURRENT_VERSION right?  It's the first time I used that guy.

Reply:
ok, done

Reply:
I'd be tempted to leave out the `#if defined(CLIENT_INFO_CURRENT_VERSION)` stanza. Other bits of rdpClientCon.c assume it's defined, particularly the version check.

Alternatively, we could add this at the start of the file if you want to run this version of xorgxrdp with xrdp versions before v0.9.16:-

```c
#ifndef CLIENT_INFO_CURRENT_VERSION
#define CLIENT_INFO_CURRENT_VERSION 0
#endif
```

We'd need to make other changes too, however, so it could turn into a bit of a rabbit hole.